### PR TITLE
Donation reassignment

### PR DESCRIPTION
**Pull request by Fareedh-12**
## Naming Rules

Name your PR like this: ISSUENUMBER-TITLE-YOURNAME

## Description

Write a summary of your change. Explain why you have made it.

## Related to

Make sure you include the issue number with a hash sign # so Github can link this PR to the right issue, like this:

Fixes #1

## Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have carefully reviewed my own code
- [ ] I have commented my code
- [ ] I have updated any documentation

### CODE REVIEW

**Comment by Fareedh-12** [comment]
@gregdyke This branch allows for donation reassignment after a laptop donation has been rejected. Please review my code. 
Pros of this implementation
> given that a request has rejected a laptop donation
> if there is another request in the db
>Then they are assigned this donation and notified

Limitations of the implementation
> it is using the number of laptops field that isn't updated. Meaning it will always reassign the donation to the number of requests equal to its number of laptops. 

Suggestions
>We start updating the field once the donation has been assigned.

File: server/api.js [hunk 1]
```diff
@@ -313,8 +313,50 @@ router.delete("/laptop_assignment/:assignmentId", async (request, response) => {
 				"(SELECT * FROM laptop_donation d WHERE d.id > $1 and (SELECT COUNT(*) FROM laptop_assignment a WHERE a.laptop_donation_id = d.id) < d.number_of_laptops ORDER BY d.id LIMIT 1)",
 				[queryResult.rows[0].laptop_donation_id]
 			);
+			const unAssignedRequests = await db.query(
+				"SELECT id, uuid, firstname FROM laptop_request  WHERE id > $1 and id NOT IN (SELECT laptop_request_id FROM laptop_assignment) and laptop_request_status != 'CANCELLED' ",
```
Comment (shieldo) [inline server/api.js:321]: I'd really like to see an explicit `ORDER BY` clause here (like in the `nextDonation` query) - while de facto a database engine is likely to fetch rows out oldest first, that isn't guaranteed to be the case, and when the order of a result set is important it's always good to explicitly think about that and include it within the query.

File: server/api.js [hunk 2]
```diff
@@ -312,8 +313,50 @@ router.delete("/laptop_assignment/:assignmentId", async (request, response) => {
...
+				[queryResult.rows[0].laptop_request_id]
+			);
+
+			// assign next request to the rejected donation
+			if (
+				unAssignedRequests.rows != undefined &&
+				unAssignedRequests.rows.length > 0
+			) {
+				let getNumberOfLaptops = await db.query(
+					"SELECT number_of_laptops FROM laptop_donation WHERE id = $1",
```
Comment (gregdyke) [inline server/api.js]: If you have a donation get unassigned, this only ever frees up 1 donation to be reassigned, so numberOfLaptops is wrong here

**Review by shieldo — COMMENTED** [review]
This seems reasonable, given the limitations which have been correctly noted.

I think it's important to think about how the logic for the endpoint here might be starting to get a little heavy - it's generally good to avoid what's often called "fat controllers" (meaning endpoints defined with complex implementations that stray from their original intent, in this case being to delete a laptop assignment (we're adding a whole number of other, quite specific things that now happen as a result of that)). It might be that other controllers (functions handling endpoints) are starting to get similarly heavy, and some refactoring may become advantageous. However, as with most things in coding, that comes with trade-offs! This is fine just now - it's good though to think about the direction a codebase might be going in, so you can best keep its complexity manageable.

I think this PR is mergeable but I'd encourage looking at adding am `ORDER BY` class to the query that's been added as per the comment.

**Review by gregdyke — CHANGES_REQUESTED** [review]
Once you have the query for the unassigned Laptops and unassigned donations, there is some shared code between POST /api/laptop_request/donation and DELETE /api/laptop_assignment.

I believe you also have a missing socket notification for the reassigned request (this socket notification does not need to happen for the POST /api/laptop_request, so that part is not shared)

File: server/api.js (hunk 2 above)
Comment (Fareedh-12) [reply server/api.js]: Oh yes, that makes clears it up.

Comment (Fareedh-12) [reply server/api.js]: I have implemented this and it still works as expected.

File: server/api.js (hunk 1 above)
Comment (Fareedh-12) [reply server/api.js:321]: I am now explicitly ordering the data from the query

**Review by gregdyke — APPROVED** [review]
The function you pulled out is less than I have in mind

File: server/api.js (hunk 1 above)
Comment (shieldo) [reply server/api.js:321]: I can still see the query

```sql
SELECT id, uuid, firstname FROM laptop_request  WHERE id > $1 and id NOT IN (SELECT laptop_request_id FROM laptop_assignment) and laptop_request_status != 'CANCELLED'
```

which doesn't have an `ORDER BY` clause. To be clear, not a blocker for merging!

**Pull request merged**